File: src/Muflone/Messages/Events/Event.cs
using MassTransit;
using Muflone.Core;
using Muflone.CustomTypes;
using System;
using System.Collections.Generic;

namespace Muflone.Messages.Events;

public class Event : IEvent
{
	public IDomainId AggregateId { get; init; }
	public EventHeaders Headers { get; set; }

	public int Version { get; set; }
	public Guid MessageId { get; set; }
	public Dictionary<string, object> UserProperties { get; set; }

	protected Event(IDomainId aggregateId, Guid correlationId, Account who)
		: this(aggregateId, correlationId, who, new When(DateTime.UtcNow))
	{
	}

	protected Event(IDomainId aggregateId, Account who)
		: this(aggregateId, Guid.Empty, who, new When(DateTime.UtcNow))
	{
	}


	protected Event(IDomainId aggregateId)
		: this(aggregateId, Guid.Empty, new Account(Guid.Empty.ToString(), "Anonymous"), new When(DateTime.UtcNow))
	{
	}

	protected Event(IDomainId aggregateId, Guid correlationId)
		: this(aggregateId, correlationId, new Account(Guid.Empty.ToString(), "Anonymous"), new When(DateTime.UtcNow))
	{
	}

	protected Event(IDomainId aggregateId, Guid correlationId, Account who, When when)
	{
		Headers = new EventHeaders
		{
			CorrelationId = correlationId,
			AggregateType = GetType().Name,
			Who = who,
			When = when
		};
		//TODO: Delete Headers. Replace with Userprops and move them in Envelope away from here
		UserProperties = new Dictionary<string, object>
		{
			{ HeadersNames.CorrelationId, correlationId }
		};

		MessageId = NewId.NextGuid();
		AggregateId = aggregateId;
	}
}

File: php/src/Muflone/Messages/Events/Event.php
<?php

declare(strict_types=1);

namespace Muflone\Messages\Events;

use DateTimeImmutable;
use Muflone\Core\IDomainId;
use Muflone\CustomTypes\Account;
use Muflone\CustomTypes\When;
use Muflone\Messages\HeadersNames;
use Ramsey\Uuid\Uuid;

class Event implements IEvent
{
    protected IDomainId $aggregateId;
    protected EventHeaders $headers;
    protected int $version = 0;
    protected string $messageId;
    /** @var array<string, mixed> */
    protected array $userProperties = [];

    public function __construct(
        IDomainId $aggregateId,
        string $correlationId = '',
        ?Account $who = null,
        ?When $when = null
    ) {
        $this->aggregateId = $aggregateId;
        $this->messageId = Uuid::uuid4()->toString();

        $who = $who ?? new Account('', 'Anonymous');
        $when = $when ?? new When(new DateTimeImmutable());

        $this->headers = new EventHeaders();
        $this->headers->setCorrelationId($correlationId);
        $this->headers->setAggregateType(static::class);
        $this->headers->setWho($who);
        $this->headers->setWhen($when);

        $this->userProperties = [
            HeadersNames::CORRELATION_ID => $correlationId,
        ];
    }

    public function getAggregateId(): IDomainId
    {
        return $this->aggregateId;
    }

    public function getHeaders(): EventHeaders
    {
        return $this->headers;
    }

    public function setHeaders(EventHeaders $headers): void
    {
        $this->headers = $headers;
    }

    public function getVersion(): int
    {
        return $this->version;
    }

    public function setVersion(int $version): void
    {
        $this->version = $version;
    }

    public function getMessageId(): string
    {
        return $this->messageId;
    }

    public function setMessageId(string $messageId): void
    {
        $this->messageId = $messageId;
    }

    /**
     * @return array<string, mixed>
     */
    public function getUserProperties(): array
    {
        return $this->userProperties;
    }

    /**
     * @param array<string, mixed> $userProperties
     */
    public function setUserProperties(array $userProperties): void
    {
        $this->userProperties = $userProperties;
    }
}
